File: client/java/src/avt/Drop_Part.java
package avt;

import javax.microedition.lcdui.Graphics;

public final class Drop_Part extends Base {
   public short idDrop;
   private short deltaH;
   public int x0;
   public int y0;
   public int ID;
   private byte g_;
   private byte dir;
   private byte i;
   private byte j = 1;
   public byte type;
   private byte state;

   public Drop_Part() {
      super.catagory = 5;
   }

   public Drop_Part(byte var1, short var2, int var3) {
      this.ID = var3;
      super.catagory = 5;
      this.type = var1;
      this.idDrop = var2;
      this.dir = 0;
      Part var4;
      if ((var4 = AvatarData.getPart(this.idDrop)).idIcon < AvatarData.listImgInfo.length) {
         super.height = AvatarData.listImgInfo[var4.idIcon].h;
      }

      this.i = (byte)CRes.rnd(10);
   }

   public final void update() {
      switch (this.state) {
         case 0:
         case 1:
            super.x += (short)(this.x0 - super.x >> 2);
            super.y += (short)(this.y0 - super.y >> 2);
            if (this.g_ >= -6) {
               this.deltaH = (short)(this.deltaH + this.g_);
               --this.g_;
            }

            if ((CRes.abs(super.x - this.x0) < 4 || CRes.abs(super.y - this.y0) < 4) && this.deltaH <= 1) {
               super.x = this.x0;
               super.y = this.y0;
               this.deltaH = 0;
               this.g_ = 0;
               if (this.state == 1) {
                  LoadMap.removePlayer((MyObject)this);
               }

               this.state = 2;
               return;
            }
         case 2:
         default:
            break;
         case 3:
            this.deltaH = (short)(this.deltaH + 3);
            if (this.deltaH > 50) {
               LoadMap.removePlayer((MyObject)this);
               return;
            }
            break;
         case 4:
            if (this.deltaH > 0) {
               this.deltaH = (short)(this.deltaH - this.g_);
               ++this.g_;
               return;
            }

            this.deltaH = 0;
            this.state = 2;
      }

   }

   public final void paint(Graphics var1) {
      var1.drawImage(LoadMap.imgShadow, super.x, super.y + 1, 33);
      if (this.type == 0) {
         AvatarData.getPart(this.idDrop).paintIcon(var1, super.x, super.y + this.i / 10 - this.deltaH, 0, 33);
      } else {
         super.height = (short)(AvatarData.getImgIcon(this.idDrop).h + 10);
         AvatarData.paintImg(var1, this.idDrop, super.x, super.y + this.i / 10 - this.deltaH, 33);
      }

      this.i += this.j;
      if (CRes.abs(this.i) >= 10) {
         this.dir = (byte)(-this.dir);
      }

   }

   public final void startFlyTo(int var1) {
      Avatar var2;
      if ((var2 = LoadMap.getAvatar(var1)) != null) {
         this.x0 = var2.x;
         this.y0 = var2.y;
         this.state = 1;
         this.deltaH = 0;
      } else {
         this.deltaH = 0;
         this.state = 3;
      }

      this.g_ = 6;
   }

   public final void startDropFrom(int var1, short var2, short var3) {
      if (var1 == -2) {
         super.x = var2;
         super.y = var3;
         this.state = 2;
      } else {
         Avatar var4;
         if ((var4 = LoadMap.getAvatar(var1)) != null) {
            super.x = var4.x;
            super.y = var4.y;
            this.state = 0;
            this.g_ = 6;
            this.deltaH = 0;
         } else {
            this.state = 4;
            super.x = var2;
            super.y = var3;
            this.deltaH = 100;
            this.g_ = 0;
         }
      }

      this.x0 = var2;
      this.y0 = var3;
   }
}

File: client/unity/Assets/Scripts/Assembly-CSharp/Drop_Part.cs
public class Drop_Part : Base
{
	public short idDrop;

	public short deltaH;

	public int x0;

	public int y0;

	public int ID;

	private new sbyte g;

	private sbyte dir;

	private sbyte yDrop;

	private sbyte yRung;

	public sbyte type;

	public sbyte state;

	public const sbyte STATE_DROP = 0;

	public const sbyte STATE_FLY = 1;

	public const sbyte STATE_STAND = 2;

	public const sbyte STATE_FLY_UP = 3;

	public const sbyte STATE_DOWN = 4;

	public const sbyte DROP_POWER = 6;

	public Drop_Part()
	{
		catagory = 5;
	}

	public Drop_Part(sbyte typeDrop, short idDrop1, int id)
	{
		ID = id;
		catagory = 5;
		type = typeDrop;
		idDrop = idDrop1;
		dir = 0;
	}

	public override void update()
	{
		switch (state)
		{
		case 0:
		case 1:
			x += (short)(x0 - x >> 2);
			y += (short)(y0 - y >> 2);
			if (g >= -6)
			{
				deltaH += g;
				g--;
			}
			if ((CRes.abs(x - x0) < 4 || CRes.abs(y - y0) < 4) && deltaH <= 1)
			{
				x = x0;
				y = y0;
				deltaH = 0;
				g = 0;
				if (state == 1)
				{
					LoadMap.removePlayer(this);
				}
				state = 2;
			}
			break;
		case 3:
			deltaH += 3;
			if (deltaH > 50)
			{
				LoadMap.removePlayer(this);
			}
			break;
		case 4:
			if (deltaH > 0)
			{
				deltaH -= g;
				g++;
			}
			else
			{
				deltaH = 0;
				state = 2;
			}
			break;
		case 2:
			break;
		}
	}

	public override void paint(MyGraphics g)
	{
		g.drawImage(LoadMap.imgShadow, x, y + 1, MyGraphics.BOTTOM | MyGraphics.HCENTER);
		if (type == 0)
		{
			Part part = AvatarData.getPart(idDrop);
			part.paintIcon(g, x, y + yRung / 20 - deltaH, 0, MyGraphics.BOTTOM | MyGraphics.HCENTER);
		}
		else
		{
			AvatarData.paintImg(g, idDrop, x, y + yRung / 20 - deltaH, MyGraphics.BOTTOM | MyGraphics.HCENTER);
		}
		yRung -= 2;
		if (yRung < -20)
		{
			yRung = 0;
		}
	}

	public void startFlyTo(int idUser)
	{
		Avatar avatar = LoadMap.getAvatar(idUser);
		if (avatar != null)
		{
			x0 = avatar.x;
			y0 = avatar.y;
			state = 1;
			deltaH = 0;
		}
		else
		{
			deltaH = 0;
			state = 3;
		}
		g = 6;
	}

	public void startDropFrom(int idPlayer, short xTo, short yTo)
	{
		if (idPlayer == -2)
		{
			x = xTo;
			y = yTo;
			state = 2;
		}
		else
		{
			Avatar avatar = LoadMap.getAvatar(idPlayer);
			if (avatar != null)
			{
				x = avatar.x;
				y = avatar.y;
				state = 0;
				g = 6;
				deltaH = 0;
			}
			else
			{
				state = 4;
				x = xTo;
				y = yTo;
				deltaH = 100;
				g = 0;
			}
		}
		x0 = xTo;
		y0 = yTo;
	}
}
